falcon_policy/middleware.py

```diff
@@ -11,13 +11,23 @@ def __init__(self, config_dict):
 
     def process_resource(self, req, resp, resource, params):
         route = req.uri_template
-        roles_header = req.get_header('X-Roles', default='@unknown')
-        provided_roles = [role.strip() for role in roles_header.split(',')]
+
+        if issubclass(req.context_type, dict):
+            roles = req.context.get('roles')
+
+        # Assume it's a direct attribute
+        else:
+            roles = req.context.roles
```

Comment:
I'm probably getting too picky, but note that the behaviour became a bit asymmetric now for *roles* from a ``dict``, and *roles* from an object.

- In the dict case, the context does not need to contain *roles* at all, triggering the fallback path, and so on.
- In the object case, missing roles attribute will cause ``AttributeError``, which is probably not a big issue now considering that people proactively choosing to use both a custom context type, and this library, are well aware of the implications. But it might become an issue should we pursue the path suggested in https://github.com/falconry/falcon/issues/1120 .

Should we consider something like ``hasattr`` on the object first?

Reply:
Generally speaking, `hasattr` should be avoided. I will however, use `getattr`.